torch/_inductor/pcache.py

```diff
@@ -0,0 +1,317 @@
+        """
+        return executor.submit(self.insert, key, value)
+
+
+class OnDiskCache(AsyncCache):
+    """
+    Abstract base class for on-disk cache implementations.
+
+    Provides synchronous and asynchronous get/insert methods for storing and retrieving data on disk.
+    All methods are thread-safe.
+
+    Methods:
+        get(key): Retrieve a value by key from disk.
+        get_async(key, executor): Asynchronously retrieve a value by key from disk.
+        insert(key, value): Insert a value on disk.
+        insert_async(key, value, executor): Asynchronously insert a value on disk.
+    """
+
+    @property
+    def base_dir(self: Self) -> Path:
+        """
+        Get the base directory for the on-disk cache.
+
+        Returns:
+            Path: The base directory for the on-disk cache.
+
+        Raises:
+            NotImplementedError: If not implemented by a subclass.
+        """
+        raise NotImplementedError
+
+    def _fpath_from_key(self: Self, key: Key) -> Path:
+        """
+        Get the file path associated with the given key.
+
+        Args:
+            key (Key): The key used to query the cache.
+
+        Returns:
+            Path: The file path associated with the key.
+        """
+        return self.base_dir / key
+
+    def _flock_from_fpath(self: Self, fpath: Path) -> FileLock:
+        """
+        Get the file lock associated with the given file path.
+
+        Args:
+            fpath (Path): The file path to lock.
+
+        Returns:
+            FileLock: The file lock associated with the file path.
+        """
+        return FileLock(str(fpath) + ".lock")
+
+    def get(self: Self, key: Key) -> Value | None:
+        """
+        Retrieve the value associated with the given key from the cache on disk.
+
+        Args:
+            key (Key): The key used to query the cache.
+
+        Returns:
+            Value | None: The value associated with the key, or None if not found.
+        """
+        fpath = self._fpath_from_key(key)
+        flock = self._flock_from_fpath(fpath)
+        with flock:
+            return fpath.read_bytes() if fpath.is_file() else None
+
+    def insert(self: Self, key: Key, value: Value) -> bool:
+        """
+        Store the given value in the cache with the associated key on disk.
+
+        Args:
+            key (Key): The key to associate with the value.
+            value (Value): The value to be stored in the cache.
+
+        Returns:
+            bool: True if the value was stored successfully, False if the key already exists.
+        """
+        fpath = self._fpath_from_key(key)
+        flock = self._flock_from_fpath(fpath)
+        fpath.parent.mkdir(parents=True, exist_ok=True)
+        try:
+            with flock:
+                # "x" mode is exclusive creation, meaning the file will be created
+                # iff the file does not already exist (atomic w/o overwrite)
+                with open(fpath, "xb") as fp:
```

Comment:
Have both with contexts on one line to save indentation and improve readability that show they are the same scope

Reply:
Done